cpp/src/lists/contains.cu

```diff
@@ -32,308 +30,256 @@
+namespace cudf::lists {
 
 namespace {
 
-auto constexpr absent_index = size_type{-1};
+/**
+ * @brief A sentinel value used for marking that a given key has not been found in the search list.
+ *
+ * The value should be `-1` as indicated in the public API documentation.
+ */
+auto constexpr __device__ NOT_FOUND_SENTINEL = size_type{-1};
 
-auto get_search_keys_device_iterable_view(cudf::column_view const& search_keys,
-                                          rmm::cuda_stream_view stream)
-{
-  return column_device_view::create(search_keys, stream);
-}
+/**
+ * @brief A sentinel value used for marking that a given output row should be null.
+ */
+auto constexpr __device__ NULL_SENTINEL = std::numeric_limits<size_type>::min();
 
-auto get_search_keys_device_iterable_view(cudf::scalar const& search_key, rmm::cuda_stream_view)
+/**
+ * @brief Indicate the current supported types in `cudf::lists::contains`.
+ *
+ * TODO: Add supported nested types.
+ */
+template <typename Element>
+static auto constexpr is_supported_non_nested_type()
 {
-  return &search_key;
+  return cudf::is_fixed_width<Element>() || std::is_same_v<Element, cudf::string_view>;
 }
 
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_begin(list_device_view const& list)
-{
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_begin<ElementType>();
-  } else {
-    return thrust::make_reverse_iterator(list.pair_rep_end<ElementType>());
-  }
-}
+/**
+ * @brief Functor to perform searching for index of a key element in a given list.
+ */
+struct search_list_fn {
+  duplicate_find_option const find_option;
 
-template <typename ElementType, duplicate_find_option find_option>
-auto __device__ find_end(list_device_view const& list)
-{
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return list.pair_rep_end<ElementType>();
-  } else {
-    return thrust::make_reverse_iterator(list.pair_rep_begin<ElementType>());
-  }
-}
+  template <typename Element, CUDF_ENABLE_IF(is_supported_non_nested_type<Element>())>
+  __device__ size_type operator()(list_device_view list, thrust::optional<Element> key_opt) const
+  {
+    // A null list never contains any key, even null key.
+    // In addition, a null list will result in a null output row.
+    if (list.is_null()) { return NULL_SENTINEL; }
 
-template <duplicate_find_option find_option, typename Iterator>
-size_type __device__ distance([[maybe_unused]] Iterator begin, Iterator end, Iterator find_iter)
-{
-  if (find_iter == end) {
-    return absent_index;  // Not found.
-  }
+    // A null key will also result in a null output row.
+    if (!key_opt) { return NULL_SENTINEL; }
 
-  if constexpr (find_option == duplicate_find_option::FIND_FIRST) {
-    return find_iter - begin;  // Distance of find_position from begin.
-  } else {
-    return end - find_iter - 1;  // Distance of find_position from end.
+    auto const& key = key_opt.value();
```

Comment:
Also, this seems like an unnecessary local variable instead of just doing:
```
  return find_option == duplicate_find_option::FIND_FIRST
             ? search_list<true, Element>(list, *key_opt)
             : search_list<false, Element>(list, *key_opt);
```